kubernetes/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -0,0 +1,222 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler.cluster.kubernetes
+
+import collection.JavaConverters._
+import io.fabric8.kubernetes.api.model.PodBuilder
+import io.fabric8.kubernetes.api.model.extensions.JobBuilder
+import io.fabric8.kubernetes.client.{ConfigBuilder, DefaultKubernetesClient}
+import org.apache.spark.internal.config._
+import org.apache.spark.scheduler._
+import org.apache.spark.scheduler.cluster._
+import org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages._
+import org.apache.spark.{SparkConf, SparkContext, SparkException}
+import org.apache.spark.rpc.RpcEndpointAddress
+import org.apache.spark.scheduler.TaskSchedulerImpl
+import org.apache.spark.util.Utils
+
+import scala.collection.mutable
+import scala.util.Random
+import scala.concurrent.Future
+
+private[spark] class KubernetesClusterSchedulerBackend(
+                                                  scheduler: TaskSchedulerImpl,
+                                                  sc: SparkContext)
+  extends CoarseGrainedSchedulerBackend(scheduler, sc.env.rpcEnv) {
+
+  val client = new DefaultKubernetesClient()
+
+  val DEFAULT_NUMBER_EXECUTORS = 2
+  val sparkExecutorName = s"spark-executor-${Random.alphanumeric take 5 mkString("")}".toLowerCase()
+
+  // TODO: do these need mutex guarding?
+  // key is executor id, value is pod name
+  var executorToPod = mutable.Map.empty[String, String] // active executors
+  var shutdownToPod = mutable.Map.empty[String, String] // pending shutdown
+  var executorID = 0
+
+  val sparkImage = conf.get("spark.kubernetes.sparkImage")
+  val clientJarUri = conf.get("spark.executor.jar")
+  val ns = conf.get(
+    "spark.kubernetes.namespace",
+    KubernetesClusterScheduler.defaultNameSpace)
+  val dynamicExecutors = Utils.isDynamicAllocationEnabled(conf)
+
+  // executor back-ends take their configuration this way
+  if (dynamicExecutors) {
+    conf.setExecutorEnv("spark.dynamicAllocation.enabled", "true")
+    conf.setExecutorEnv("spark.shuffle.service.enabled", "true")
+  }
+
+  override def start(): Unit = {
+    super.start()
+    createExecutorPods(getInitialTargetExecutorNumber(sc.getConf))
+  }
+
+  override def stop(): Unit = {
+    // Kill all executor pods indiscriminately
+    killExecutorPods(executorToPod.toVector)
+    killExecutorPods(shutdownToPod.toVector)
+    super.stop()
+  }
+
+  // Dynamic allocation interfaces
+  override def doRequestTotalExecutors(requestedTotal: Int): Future[Boolean] = {
+    logInfo(s"Received doRequestTotalExecutors: $requestedTotal")
+    val n = executorToPod.size
+    val delta = requestedTotal - n
+    if (delta > 0) {
+      logInfo(s"Adding $delta new executors")
+      createExecutorPods(delta)
+    } else if (delta < 0) {
+      val d = -delta
```

Comment:
Delta (as currently computed) can become negative, whenever the requested total starts to decrease.

However, this logic warrants a total overhaul, since it was designed around the assumption that `doRequestTotalExecutors` is symmetric, in the sense of having to handle a decreasing total analogously to an increasing total.  But responsibilities for executor shutdown vs spin-up aren't symmetric.  I might file a small doc PR to clarify that point.